services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/crt/CopyHelper.java

```diff
@@ -0,0 +1,248 @@
+            if (throwable != null) {
+                handleException(returnFuture, () -> "Failed to retrieve metadata from the source object", throwable);
+            } else {
+                Long contentLength = headObjectResponse.contentLength();
+
+                long partSize = s3NativeClientConfiguration.partSizeBytes();
+
+                if (contentLength <= partSize) {
+                    log.debug(() -> "Start the copy as a single copy part requests");
+                    copyInOneChunk(copyObjectRequest, returnFuture);
+                } else {
+                    log.debug(() -> "Start the copy as multiple copy part requests");
+                    copyInParts(copyObjectRequest, contentLength, returnFuture);
+                }
+            }
+        });
+
+        return returnFuture;
+    }
+
+    private void copyInParts(CopyObjectRequest copyObjectRequest,
+                             Long contentLength,
+                             CompletableFuture<CopyObjectResponse> returnFuture) {
+
+        CreateMultipartUploadRequest request = CopyRequestConversionUtils.toCreateMultipartUploadRequest(copyObjectRequest);
+        CompletableFuture<CreateMultipartUploadResponse> createMultipartUploadFuture = s3AsyncClient.createMultipartUpload(request);
+
+        // Ensure cancellations are forwarded to the createMultipartUploadFuture future
+        CompletableFutureUtils.forwardExceptionTo(returnFuture, createMultipartUploadFuture);
+
+        createMultipartUploadFuture.whenComplete((createMultipartUploadResponse, throwable) -> {
+            if (throwable != null) {
+                handleException(returnFuture, () -> "Failed to initiate multipart upload", throwable);
+            } else {
+                log.debug(() -> "Initiated new multipart upload, uploadId: " + createMultipartUploadResponse.uploadId());
+                doCopyInParts(copyObjectRequest, contentLength, returnFuture, createMultipartUploadResponse.uploadId());
+            }
+        });
+    }
+
+    private int determinePartCount(long contentLength, long partSize) {
+        return (int) Math.ceil(contentLength / (double) partSize);
+    }
+
+    private void doCopyInParts(CopyObjectRequest copyObjectRequest,
+                               Long contentLength,
+                               CompletableFuture<CopyObjectResponse> returnFuture,
+                               String uploadId) {
+        long optimalPartSize = calculateOptimalPartSizeForCopy(s3NativeClientConfiguration.partSizeBytes());
+
+        int partCount = determinePartCount(contentLength, optimalPartSize);
+
+        log.debug(() -> String.format("Starting multipart copy with partCount: %s, optimalPartSize: %s",
+                                      partCount, optimalPartSize));
+
+        // The list of completed parts must be sorted
+        AtomicReferenceArray<CompletedPart> completedParts = new AtomicReferenceArray<>(partCount);
```

Comment:
Implementation note: the reason `AtomicReferenceArray` is used instead of `ConcurrentLinkedQueue` is that the completedParts must be ordered and `AtomicReferenceArray` should be more performant in this case